### PR TITLE
support loading gzipped tiles

### PR DESCRIPTION
**Pull request by noblige**

### CODE REVIEW

File: valhalla/baldr/graphtile.h [hunk 1]
```diff
@@ -315,7 +315,7 @@ class GraphTile {
 
   // Graph tile memory, this must be shared so that we can put it into cache
   // Apparently you can std::move a non-copyable
-  boost::shared_array<char> graphtile_;
+  boost::shared_ptr<std::vector<char>> graphtile_;
```
Comment (kevinkreiser) [inline valhalla/baldr/graphtile.h]: having it as a `vector` maybe we dont even need it to be a `shared_ptr` we can essentially just change anything that does:

    if(graphtile_)

into:

    if(graphtile_.size())

what do you think?

Comment (kevinkreiser) [inline valhalla/baldr/graphtile.h]: wondering if this can just be `std::vector<char> graphtile_;`

Comment (kevinkreiser) [inline valhalla/baldr/graphtile.h]: i think we can only do that if we make `graphtile` non copyable which might be a good idea. maybe i'll try that later after we get this merged!

**Comment by kevinkreiser** [comment]
this looks good ill test it locally

File: valhalla/baldr/graphtile.h (hunk 1 above)
Comment (noblige) [inline valhalla/baldr/graphtile.h]: I think replacing this with `boost::scoped_ptr` will make it non-copyable. or as you said we can just use plain vector and make sure `GraphTile` itself non-copyable.

**Comment by noblige** [comment]
Hm, for some reason this does not link on ARM, looks like `libtool` puts libraries in a different order. Trying to sort this out:
```
  CXXLD  valhalla_run_map_match
./.libs/libvalhalla.so: undefined reference to `boost::iostreams::detail::gzip_footer::reset()'
...
```

**Comment by kevinkreiser** [comment]
im not sure i understand whats wrong but i noticed you put it on the end of all the lines in the makefile. unrelated to this, i was actually wondering if we might do something like this to make changes like this to make managing boost easier for us #511

**Comment by noblige** [comment]
@kevinkreiser looking at #511 for some reason I get same linking errors on x86 _unless_ I put `iostreams` at the end. which is weird

**Comment by noblige** [comment]
but for arm, it turns out I've compiled boost without zlib support :|

**Comment by noblige** [comment]
once #511 is merged, I can rebase and kick off a build again - maybe it will magically work!

**Comment by noblige** [comment]
@kevinkreiser thanks. I will investigate more, but it appears that something is wrong with tile cache, as I can see same tile being constructed over and over again (and with them being gzipped  is super slow)

**Comment by kevinkreiser** [comment]
@noblige that will happen if you set the tile cache size really low. this value: https://github.com/valhalla/conf/blob/master/valhalla.json#L3 if you set that really low each request could in theory kick everything out of cache only to re-load it. if that is set moderately high then it shouldn't happen. let me take a quick look

**Comment by noblige** [comment]
@kevinkreiser yeah, I bumped it 5G, just to make sure - still same issue. thanks!

**Comment by kevinkreiser** [comment]
what the heck! ok just to make sure im understanding, you're saying no matter what setup used (gzipped or not) it shows over and over not hitting cache but calling the `GraphTile` constructor?

**Comment by kevinkreiser** [comment]
this is what i get using gzipped tiles:

```
2017/02/13 18:20:56.834811 [33;1m[WARN][0m Tile extract could not be loaded
0 2017/02/13 18:21:08.299984 GET /route?json={%22locations%22:[{%22lat%22:40.35897186956936,%22lon%22:-76.42740726492775,%22type%22:%22break%22},{%22lat%22:40.34286441417128,%22lon%22:-76.41454696655273,%22type%22:%22through%22,%22name%22:%22%22},{%22lat%22:40.33991232633096,%22lon%22:-76.41300201416016,%22type%22:%22break%22,%22name%22:%22%22}],%22costing%22:%22auto%22,%22directions_options%22:{%22language%22:%22en-US%22}}&api_key= HTTP/1.1
2017/02/13 18:21:08.300168 [32;1m[INFO][0m Got Loki Request 0
2017/02/13 18:21:08.301347 [ANALYTICS] location_count::3
2017/02/13 18:21:08.301381 [ANALYTICS] costing_type::auto
2017/02/13 18:21:08.301709 [ANALYTICS] location_distance::2.098873km
2017/02/13 18:21:08.301727 [ANALYTICS] location_distance::0.353737km
2017/02/13 18:21:08.314505 [31;1m[ERROR][0m LOADING 6005234
2017/02/13 18:21:08.374871 [31;1m[ERROR][0m LOADING 23240
2017/02/13 18:21:08.405474 [31;1m[ERROR][0m LOADING 375225
2017/02/13 18:21:08.408899 [32;1m[INFO][0m Got Thor Request 0
2017/02/13 18:21:08.410943 [ANALYTICS] travel_mode::0
2017/02/13 18:21:08.458022 [31;1m[ERROR][0m LOADING 375225
2017/02/13 18:21:08.467780 [31;1m[ERROR][0m LOADING 6005234
2017/02/13 18:21:08.523142 [31;1m[ERROR][0m LOADING 23240
2017/02/13 18:21:08.568887 [ANALYTICS] admin_state_iso:: PA 
2017/02/13 18:21:08.568923 [ANALYTICS] admin_country_iso:: US 
2017/02/13 18:21:08.569455 [32;1m[INFO][0m Got Odin Request 0
2017/02/13 18:21:08.949835 [32;1m[INFO][0m trip_path_->node_size()=31
2017/02/13 18:21:08.955118 [32;1m[INFO][0m maneuver_count::10
2017/02/13 18:21:08.955499 [32;1m[INFO][0m Got Tyr Request 0
2017/02/13 18:21:08.957231 [ANALYTICS] language::en-US
2017/02/13 18:21:08.958115 [ANALYTICS] trip_length::5.012669km
0 2017/02/13 18:21:08.958377 200 5177
1 2017/02/13 18:21:43.523565 GET /route?json={%22locations%22:[{%22lat%22:40.35897186956936,%22lon%22:-76.42740726492775,%22type%22:%22break%22},{%22lat%22:40.34286441417128,%22lon%22:-76.41454696655273,%22type%22:%22through%22,%22name%22:%22%22},{%22lat%22:40.33991232633096,%22lon%22:-76.41300201416016,%22type%22:%22break%22,%22name%22:%22%22}],%22costing%22:%22auto%22,%22directions_options%22:{%22language%22:%22en-US%22}}&api_key= HTTP/1.1
2017/02/13 18:21:43.523734 [32;1m[INFO][0m Got Loki Request 1
2017/02/13 18:21:43.524456 [ANALYTICS] location_count::3
2017/02/13 18:21:43.524478 [ANALYTICS] costing_type::auto
2017/02/13 18:21:43.524597 [ANALYTICS] location_distance::2.098873km
2017/02/13 18:21:43.524609 [ANALYTICS] location_distance::0.353737km
2017/02/13 18:21:43.530578 [32;1m[INFO][0m Got Thor Request 1
2017/02/13 18:21:43.533534 [ANALYTICS] travel_mode::0
2017/02/13 18:21:43.572251 [ANALYTICS] admin_state_iso:: PA 
2017/02/13 18:21:43.572287 [ANALYTICS] admin_country_iso:: US 
2017/02/13 18:21:43.572621 [32;1m[INFO][0m Got Odin Request 1
2017/02/13 18:21:43.574368 [32;1m[INFO][0m trip_path_->node_size()=31
2017/02/13 18:21:43.579240 [32;1m[INFO][0m maneuver_count::10
2017/02/13 18:21:43.579491 [32;1m[INFO][0m Got Tyr Request 1
2017/02/13 18:21:43.581659 [ANALYTICS] language::en-US
2017/02/13 18:21:43.582358 [ANALYTICS] trip_length::5.012669km
1 2017/02/13 18:21:43.582479 200 5177
```

The second request doesnt have any messages about loading tiles. I added that message in the `GraphTile::Initialize`.

**Comment by noblige** [comment]
@kevinkreiser For some reason this only happens on ARM. I'm at loss here, but will dig deeper.

**Comment by kevinkreiser** [comment]
wild! yeah please let us know what you find out!

**Comment by noblige** [comment]
@kevinkreiser was accidentally running with more than 1 thread :| With 1 thread works as expected, however, for 1st request I'm still seeing same tile loaded more than once:

```
2017/02/13 11:29:09.450941 [INFO] GraphReader::GraphReader
2017/02/13 11:29:09.455273 [INFO] GraphReader::GraphReader
2017/02/13 11:29:09.457944 [INFO] GraphReader::GraphReader
2017/02/13 11:29:09.492052 [INFO] GraphReader::GraphReader
2017/02/13 11:29:12.320361 [INFO] Got Loki Request 0
2017/02/13 11:29:12.321493 [ANALYTICS] location_count::2
2017/02/13 11:29:12.321650 [ANALYTICS] costing_type::auto
2017/02/13 11:29:12.322227 [ANALYTICS] location_distance::57.350330km
2017/02/13 11:29:12.958647 [INFO] Initialize: 5877042
2017/02/13 11:29:13.948340 [INFO] Initialize: 5865538
2017/02/13 11:29:15.152285 [INFO] Initialize: 22432
2017/02/13 11:29:16.661444 [INFO] Initialize: 366225
2017/02/13 11:29:18.567989 [INFO] Initialize: 366217
2017/02/13 11:29:18.572479 [INFO] loki::request elapsed time (ms)::6252.076660
2017/02/13 11:29:18.573473 [INFO] GraphReader::OverCommitted: 50000000 < 49646602
2017/02/13 11:29:18.573547 [INFO] Got Thor Request 0
2017/02/13 11:29:18.575543 [ANALYTICS] travel_mode::0
2017/02/13 11:29:19.875479 [INFO] Initialize: 5865538
2017/02/13 11:29:21.850175 [INFO] Initialize: 366217
2017/02/13 11:29:23.240958 [INFO] Initialize: 366225
2017/02/13 11:29:23.844217 [INFO] Initialize: 5877042
2017/02/13 11:29:25.049607 [INFO] Initialize: 22432
2017/02/13 11:29:25.690672 [INFO] path_cost::3378
2017/02/13 11:29:25.690887 [INFO] FormPath path_iterations::8538,13505
2017/02/13 11:29:25.810052 [ANALYTICS] admin_state_iso:: CA 
2017/02/13 11:29:25.810403 [ANALYTICS] admin_country_iso:: US 
2017/02/13 11:29:25.819989 [INFO] thor::route trip_path elapsed time (ms)::7244.248047
2017/02/13 11:29:25.824982 [INFO] Got Odin Request 0
2017/02/13 11:29:25.833378 [INFO] trip_path_->node_size()=409
2017/02/13 11:29:25.842518 [INFO] EXIT RelativeDirection=1
2017/02/13 11:29:25.843014 [INFO] EXIT RelativeDirection=1
2017/02/13 11:29:25.914486 [INFO] GraphReader::OverCommitted: 50000000 < 0
2017/02/13 11:29:25.921918 [INFO] GraphReader::OverCommitted: 50000000 < 49646602
2017/02/13 11:29:26.144105 [INFO] maneuver_count::24
2017/02/13 11:29:26.147516 [INFO] Got Tyr Request 0
2017/02/13 11:29:26.148217 [ANALYTICS] language::en-US
2017/02/13 11:29:26.155006 [ANALYTICS] trip_length::76.392975km
```

**Comment by noblige** [comment]
@kevinkreiser it is same in your example as well - is there a reason tiles are loaded twice? (once in Loki and then in Thor)

**Comment by kevinkreiser** [comment]
correct. loki and thor run in different threads, they don't share storage across the thread boundry. the only way to get around this is to use memory mapped tar configuration in valhalla.json

**Comment by noblige** [comment]
@kevinkreiser this is not great for embedded - I will work on adding thread safe `GraphReader` that can be shared between threads. How does that sound?

**Comment by kevinkreiser** [comment]
@noblige i think that might be a bit hairy! maybe you could do something with a graphreader singleton and a mutex? although when they are run in separate processes that would just added unneeded complexity (and latency).

in my opinion, though, for the purposes of embedded (where high concurrency isnt important) it would be better to run valhalla_route_service in an alternate mode that doesnt make separate threads for each stage of the pipeline. Instead it would have a single monolithic request handler that shares the same graphreader throughout.

**Comment by noblige** [comment]
@kevinkreiser I need to take a closer look how it is organized - I think problem is that GraphReader gives out raw pointers, that are obviously will get invalidated in someone calls `Clear()`
I think I get can away for my particular case (`valhalla_route_service` with 1 thread) - by making singleton instance of the cache. Or there could still be a problem with concurrency?

Here's rough idea - will try this and see what are memory/CPU savings:

https://github.com/valhalla/valhalla/compare/master...teslamotors:shared-cache?expand=1

**Pull request merged**